### ISSUE TITLE
Support locking modules

### ISSUE THREAD

**Issue by stevehipwell** [opentofu/opentofu#586]
### OpenTofu Version

```shell
ALL
```

### Use Cases

I'd like to be able to have modules locked in the same way that provider versions are locked so my IaC can be idempotent.

### Attempted Solutions

You can do this by consuming modules from a git source by SHA reference but this has none of the UX benefits to being able to consume modules from a registry or at least by a version tag.

**Comment by eranelbaz** [comment]
What about
```
module "name" {
  source  = "source"
  version = "0.0.5"
}
```
why isn't that enough?

**Comment by stevehipwell** [comment]
@eranelbaz that would require an immutable registry, an explicit module version and still wouldn't cover Git usage without referencing the SHA.

```terraform
module "hello" {
  source  = "github.com/org/repo.git?ref=v1.0.0"
}
```

I'd like to be able to do any of the following and have the version locked at the point of running `tofu init -upgrade`.

```terraform
module "hello_1" {
  source  = "source"
  version = "~> 1.0.0"
}

module "hello_2" {
  source  = "github.com/org/repo.git?ref=v1.0.0"
}

module "hello_3" {
  source  = "github.com/org/repo.git?ref=main"
}
```

**Comment by stevehipwell** [comment]
@eranelbaz a git tag isn't immutable.

**Comment by eranelbaz** [comment]
So how can we make it idempotent in this case?
The module registry referencing git tags, and if module maintainer decided to change the tag next `tofu init` will pull the new tagged version - and usually you do it to hotfix some critical issues where the feature set doesn't change

You can use git hash if you really want it to init the same version.
```
module "storage" {
  source = "git::https://example.com/storage.git?ref=51d462976d84fdea54b47d80dcabbf680badcdb8"
}
```

but

**Comment by eranelbaz** [comment]
BTW, also provider versions are based on tags

**Comment by stevehipwell** [comment]
> So how can we make it idempotent in this case?
The module registry referencing git tags, and if module maintainer decided to change the tag next tofu init will pull the new tagged version - and usually you do it to hotfix some critical issues where the feature set doesn't change

The actual module version should be idempotent even if the constraint isn't, which is how providers work. So as with providers modules should have their hash stored in the lock file.

> BTW, also provider versions are based on tags

And the lock file stops the use of a retagged provider unless the lock file is updated to reflect this.

**Comment by eranelbaz** [comment]
Oh ok, you mean for the lock file, I thought you mean for the registry that serves those modules / regsitry

**Comment by stevehipwell** [comment]
> Oh ok, you mean for the lock file, I thought you mean for the registry that serves those modules / regsitry

That is correct.

Not related but doing something similar to how Go blocks tag changes might be worth investigating in the future.

**Comment by eranelbaz** [comment]
Ok Ok
Now I'm understanding the use-case
thank you for the explanation!

**Comment by archoversight** [comment]
This would be great as well for CI/CD pipelines where pulling in an updated version cause the developer tested with an older cached version leads to errors/issues that are hard to debug because it is not immediately obvious.

**Comment by ziggythehamster** [comment]
I came to +1 this and also suggest that a new stanza in `.terraform.lock.hcl` (`.opentofu.lock.hcl`?) would be sufficient. Think like a `Gemfile.lock` except that you also cryptographcially validate that the file you fetch hasn't changed.

This could be as simple as this:

```hcl
module "gitlab.company.dev/organization/module-name/aws" "root.foo" {
  version     = "1.0.1234"
  constraints = "~> 1.0"
  hashes = [
    "h1:3pi8UzZsZ0LH4fl7cglkLKkSXPPol8cugI1iqwNw+VA=",
    "zh:001e2886dc81fc98cf17cf34c0d53cb2dae1e869464792576e11b0f34ee92f54",
  ]
}
```

The first component of the module stanza address (is there standardization on what this is called?) would be the normalized and fully qualified source address, so it would work with git sources as well.

**Comment by wonko** [comment]
you would have to keep track of which version is used for which instance of the module within the plan. Imagine a plan with
```
module "a" {
  source = "some/module/somewhere"
  version = "~> 1.0"
}

module "b" {
  source = "some/module/somewhere"
  version = "~> 2.0"

  a_new_feature = true
}
```

The lockfile would need to keep track that `module A` could currently be at p.ex. 1.3, and `module B` could be at 2.1.

For me, the main purpose is to make sure that two developers working are 100% certain that they are using exactly the same modules, and that this behaviour can be replicated exactly in a pipeline. Too many times, we've had small changes being applied and removed (some simple policy statements, a single setting...) by applying an out-of-sync plan on two developers workstations.

This is exactly what the current lockfile does for the providers. Even more, when running a new plan, it's easy to check if your installed modules are actually aligned with the lockfile, and suggest to do an upgrade/downgrade towards the recorded versions.

**Comment by davidnortonjr** [comment]
Another valuable reason to use this is related to transitive dependencies.  Modules can reference other submodules and every time you run `terraform init` you could get a different version of the sub-sub-module, depending on the version constraint. 

Consider my code `A`, which uses module `B`, which in turn uses module `C`. In `A` I can reference a specific commit sha of `B`, but `B` could reference a version range on `C` and you are at the mercy of `B` defining a sensible range, and at the mercy of `C` behaving as expected with regards to breaking changes.

By locking modules in the lockfile, you can lock all modules, including transitive dependencies -- ensuring idempotency.

**Comment by cam72cam** [comment]
RFC: Module Locking

# Summary

For an IaC tool, security and reproducability is paramount.

One of the methods to achive this in OpenTofu is Provider Locking. Providers are hashed and the results are stored within a file in the project repository. This is validated before every major operation to ensure that the expected binary is being executed.

Modules with the exec provisioner can execute arbitrary code and should be treated with the same care as providers. With the addition of provider functions, this concern becomes even more paramount.

# User Facing Description

As a user, I expect that any non-local module (registry or remote) I pull in will have a corresponding entry in the lock file. This should be checked every time OpenTofu downloads a module, and eventually when it executes a major operation.

# Technical Description

Modules can be sourced from a registry, a repository, or a remote archive. They are downloaded (and extracted if nessesary) to .terraform/modules/ in a path representing the module's name and location within the project.

There are two different use-cases for locking a module:
1. Confirming downloaded module sources are correct and have not been tampered with
2. Confirming already downloaded module sources are correct.

I believe case 1 is the more common use case and can be fairly easily accomplished without much overhead. Immediatly after a module is downloaded, gosum can be used to hash the module directory and to record the result in .terraform.lock.hcl, mirroring the provider process. Additionally, multiple modules which need the same sources will be downloaded and validated once before being copied to the different instances.

Case 2 is a bit more tricky. The first challenge with case 2 is performance, as a large project with large modules will require lots of disk and CPU resources to validate. That can be somewhat mitigated with faster disks / use of goroutines.

The second challenge with case 2 is that modules may be self-modifying. They are allowed (but discouraged) to write files into their .terraform/modules/ directory (or anywhere else on the filesystem for that matter...). Any additional files added to the module source directory will then be included in that hash, which will cause differences from the lock file. Users also may temporarily modify files in .terraform/modules, although it is discouraged.

Therefore we should focus on implementing case 1, while leaving the door open for optionally implementing case 2 (potentially at a later date).

Therefore, I propose that in the first stage of implementation, the following is done:
* Refactor the hashing out of the getproviders package to a common location so it can be used in getmodules.
* Refactor the depsfile package to include "module" declarations in the lockfile.
  - "module" declarations should either use "source" or a combination of "source" and "version reqs" as the identifier.
* Integrate the depfile validation with the getmodules package.
* Add an environment variable / `terraform/tofu` config flag to disable module caching

For the second stage, if and when we decide to implement it:
* Add module validation to all remaining locations that provider validation occurs
* Add a new env/config for enabling/disabling this additional checking.
* Warn in the documentation that modules that change the contents of their directory can not be used when module locking is enabled.


# Rationale and alternatives

An external tool may be able to be written to provide the same functionality, generating checksums of the files within the repository.  This is not ideal, as the solution should be integrated into OpenTofu directly.

# Downsides

This will add a non trivial amount of time to the `tofu init` command depending on the size and complexity of a project. 

As of 1.5.x terraform errors when encountering "module" blocks in .terraform.lock.hcl. If we add "module" blocks, it will increase the difficulty of migrating back to terraform. Tooling / flags could be added to provide a migration path.

# Unresolved Questions

How much do we care about hash collisions?

When should this be enabled by default?

**Comment by ziggythehamster** [comment]
> you would have to keep track of which version is used for which instance of the module within the plan

To clarify, that's why my proposed syntax in https://github.com/opentofu/opentofu/issues/586#issuecomment-1756350811 is `module "<source>" "<module path>" {`.

A more complete example of my proposal would look something like this:

.opentofu.lock.hcl (or any other appropriate file, maybe .terraform.module-lock.hcl):

```hcl
module "gitlab.company.dev/organization/module-name/aws" "root.foo" {
  version     = "1.0.1234"
  constraints = "~> 1.0"
  hashes = [
    "h1:3pi8UzZsZ0LH4fl7cglkLKkSXPPol8cugI1iqwNw+VA=",
    "zh:001e2886dc81fc98cf17cf34c0d53cb2dae1e869464792576e11b0f34ee92f54",
  ]
}
```

then in foo.tf:

```hcl
module "foo" {
  source  = "gitlab.company.dev/organization/module-name/aws"
  version = "~> 1.0"
}
```

The reason why I have the module path as `root.foo` is because this is the syntax used in the state file, and I assume is already easy to retrieve for any arbitrary module.

`h1` hashes contain the full hierarchy of a subtree and can be used to verify integrity in the case of self-modifying modules (or administrators poking around). [reference](https://pkg.go.dev/golang.org/x/mod/sumdb/dirhash)

**Comment by cam72cam** [comment]
Hmm, I'm curious if the "root.foo" is needed.  I don't think we need to know which module instance/call it is directly.  What we really want to know is if a given module source matches what we expect.

There are some efficiency gains related to file size and number of hashes performed during `tofu init` if it's keyed on module source and not module instance.

**Comment by wonko** [comment]
Can't you have the same module twice, with a different version? I can't think of a concrete situation, but it would make sense to allow this:

```
module "foo" {
  source  = "gitlab.company.dev/organization/module-name/aws"
  version = "1.2.3"
}

module "bar" {
  source  = "gitlab.company.dev/organization/module-name/aws"
  version = "3.4.5"
}
```
For this, you will have to keep track of `root.foo` and `root.bar` and the constraints that go with each of these.

(The main gain for me is repeatability across multiple system, no more "it runs on my machine". If you do `tofu init`, it will set to the exact same configuration as the lock file is, allowing the same across multiple execution-systems (pipelines, local system, automated systems in any way...).)

**Comment by cam72cam** [comment]
> "module" declarations should either use "source" or a combination of "source" and "version reqs" as the identifier.

Which would result in the following lock contents fom your example:
```hcl
module "gitlab.company.dev/organization/module-name/aws" "1.2.3" {
}

module "gitlab.company.dev/organization/module-name/aws" "3.4.5" {
}
```

That way other modules calls that have identical requirements could re-use the same hash lock

**Comment by ziggythehamster** [comment]
> That way other modules calls that have identical requirements could re-use the same hash lock

If you have two `~> 1.0` requirements, they can resolve to different versions, so it would be necessary to be careful and make sure that two modules with the same constraint must resolve to the exact same actual version.

**Comment by leanrobot** [comment]
Curious for a status update on this issue, please! 🙏

**Comment by apparentlymart** [comment]
Hi @leanrobot,

This issue is still on our radar, but it's relatively low in [the popular issues chart](https://github.com/opentofu/opentofu/issues/1496) and so the core team is currently focusing more on the items near the top of those lists.

In the meantime we encourage folks to continue adding 👍 reactions to the original issue to influence that prioritization. Thanks!

**Comment by joeaawad** [comment]
I have a lot of concerns that this proposal would result in a solution similar to provider locking that simply does not scale as your usage gets into the hundreds or thousands of root modules and state files. I agree that this would work great if you only have a few root modules. 

If OpenTofu wanted to tackle this challege, I would advocate for something that worked more like `tenv` that supported both provider and module locking. What I mean by this is that instead of having to have a separate lock file in each individual root module, you would also support picking up a single lockfile higher up in the directory structure (or in the home directory) that is then inherited by all of the child directories. 

Granted, this idea has a lot of uncertainty to me. Maybe you have to have a file in the same directory as the lockfile that has to be manually maintained and contains the superset of all providers or module packages that get invoked in the child modules? Maybe OpenTofu adds a new command that is able to search all child directories and generate a lockfile that is the superset of all of those? Regardless, the solution I'm envisioning here would be flexible enough to support both individual root module locking and inherited/more centralized locking so that as a company scales, they can adjust the way that they have implemented the locking to manage multiple root modules at once without having to build a bunch of custom tooling.

I've written [this document](https://github.com/joeaawad/random-scripts/blob/master/terraform-opentofu-versioning.md) for some additional background on the issues with the current provider locking system.

**Comment by stevehipwell** [comment]
@joeaawad this issue is for module locking in the context of a workspace, meaning a single lock file and a single state file.

The current provider locking implementation works well in this context as long as you're careful to distinguish between workspaces, which can have lock files, and modules, which can't have lock files. The workspace should be explicit in it's dependencies, using `~> x.y` constraints and a lock file for determinism. Modules should be permissive, using `>= x.y.z` constraints and only constraining the upper bounds (e.g. `>= x.y.z < x+1`) when absolutely required. This pattern allows for automated workspace version management and only requires additional module code changes when there is a breaking provider change that the module can't support yet. Even without the module change, the workspace can still control the versions so the workspace owner is still in full control and wont be blocked waiting on a new module version. If you add a tool like Dependabot, configure it to groups non-major changes, and have automated testing, this whole process has very little toil.

**Comment by apparentlymart** [comment]
I think it's worth noting that the status quo for module dependencies is already different from provider dependencies in an important way: 

An entire OpenTofu configuration must agree on a single version of each provider to use, and so the decision in the lock file is one made on behalf of every module in the configuration. If any one of the modules needs a newer version of a provider then that new selection applies to all of them. I can understand then that some teams would want to extend that global decision beyond a single configuration so that they can easily elect to upgrade every use of a provider across all configurations at once. (Whether that's a good idea in practice, given how existing providers tend to approach breaking changes, is not something I will wade into here.)

In contrast, today each `module` block makes its own separate decision about which module package to use, and so if we change nothing else except adding lock file entries then we'd have a separate lock file entry for each `module` block in the configuration. That model therefore _cannot_ extend "globally" (across many configurations at once) because there is currently no global address for a module call across many separate configurations.

A big part of why this issue isn't easy is that there doesn't seem to be a single shared idea about what exactly it means to track selections of module packages in the dependency lock file. Finding a suitable design that meets as many different workflow preferences as possible is going to be an interesting challenge!

**Comment by stevehipwell** [comment]
@apparentlymart isn't the current status quo effectively a "lock" per module block where the lock exists for the lifetime of the last init call (or equivalent)?

**Comment by apparentlymart** [comment]
Yes, the `.terraform/modules/modules.json` file is _essentially_ a lock file that just isn't in version control, and indeed it tracks each module call separately like I was describing.

There are some subtleties that mean that this file isn't really a lock file in the sense I think this issue intends it to mean:

- It doesn't track checksum information, so it can't notice if the module content has been modified.
- It doesn't track version selections in the same sense as we do for providers, where we expect `tofu init` to always select exactly the same version of the provider as was previously selected unless the `-upgrade` option is active.

So in practice OpenTofu currently relies on a combination of that metadata file and the content of the directories that were populated by `tofu init` in order to "remember" what was installed. If you delete the `.terraform/modules` directory, or if you run `tofu init` on a different computer where that directory doesn't already exist, then OpenTofu could end up installing entirely different module package contents.

My previous comment was mainly reacting to https://github.com/opentofu/opentofu/issues/586#issuecomment-3268758860 by pointing out that, unless we also change other details of how module package installation works, I would expect that tracking module package selections in the dependency lock file would have its own quite different set of design tradeoffs and workflow challenges to what we currently know with providers. (I don't mean to say that it would necessarily be better or worse, only that it would be different.)

**Comment by joeaawad** [comment]
What exactly is the scope of this issue? Is it only referring to locking remote module sources or is there also some consideration for local module sources? I'm assuming we are only talking about remote module sources here but we should be sure to clarify that scope.

**Comment by joeaawad** [comment]
I think that part of this, particularly when you are talking about a more global or inherited module lock, is heavily influenced by the granularity of your module packages. If you have a separate module package for each individual module, I can see an argument that a global or inherited module lock is less useful for your use case. But if you use TF at scale and have a lot of inter-connected shared modules, there are a lot of companies that have a single large module package where individual modules are then referenced as subdirectories in that single large module package to help minimize incompatibility issues between different versions of individual modules. In this use case, the module package is more like a provider and a global or inherited lockfile could be a lot more useful.

**Comment by joeaawad** [comment]
Also I want to flag https://github.com/opentofu/opentofu/issues/1199 as possibly related since that could involve introducing a more global cache for remote module packages.

**Comment by stevehipwell** [comment]
> Yes, the .terraform/modules.json file is essentially a lock file that just isn't in version control, and indeed it tracks each module call separately like I was describing.

@apparentlymart given this, a lockfile that referenced each module by path (the module identity) would map to the current behaviour and would allow portability between machines. As a MVP solution this would likely be good enough for the 80% use cases.

A more advanced version of the above pattern could use module `source` & `version` as the identity which would allow for deduplication of lock entries while keeping the upgrade changes easy to explain.

As a related subject, locking modules does make a case for allowing OCI modules to use `version` to set a version constraint instead of a fixed reference.

**Comment by tmccombs** [comment]
> the .terraform/modules.json file is essentially a lock file that just isn't in version control

The fact that it isn't in version control is a signifcant problem though. That means that CI may use a different version of a module than was used in developement, or even the last CI run, if the `.terraform` folder is not somehow preserved between CI runs.

**Comment by apparentlymart** [comment]
In my earlier comment I was focusing on the scope of the dependency lock file because that's what started the fresh discussion today, but if we want to talk about design questions about this more broadly then there's a much more significant one that still remains:

Aside from registry-based module sources, OpenTofu's module installer doesn't have any concept of versions and so it isn't really clear what kind of selection the lock file would even be capturing.

The registry protocols allow us to ask "what versions are available?" and "where is the package for version N?", and so during an "upgrade" (or when there isn't yet a selection in the lock file" we ask the first question, and then when we're reinstalling something previously selected we ask the second question in order to find the same thing we previously selected.

The non-registry sources only allow us to install whatever happens to be current at the moment we ask. There is no way to list available versions or to request the location of a version number we previously remembered. If we tried to track checksums for those without changing anything else today then it would break anyone who isn't specifying an immutable artifact (like a git commit or oci image digest) in their source address, because the checksum would fail as soon as the selected artifact changes.

This would be partially solved by something like https://github.com/opentofu/opentofu/issues/2495, or by https://github.com/opentofu/opentofu/pull/2892. We could then choose to make a compromise where source types that support version selection get tracked in the lock file but unversioned source types do not, and then those who are worried about module package integrity checks could make sure to use only source types that can support them. 

This still has a relatively small number of upvotes though, so we aren't prioritizing any detailed design for any of that. In the meantime you the best option we have available is to use the two immutable source address types where a checksum is inherent in the source address: a git source with a specific commit id, or an OCI repository source with a specific image digest.

**Comment by stevehipwell** [comment]
@apparentlymart I'm not sure I see a major blocker to locking modules as long as they're registry based or have a content address (Git or OCI). With a lock file targeting a module identity (path or combination of `source` & `version`) the resolved target (registry version or digest) could be locked alongside the actual module checksums. The new logic would then be for the module loader to use the lock file to override the target when specified, module locking and modules not locked would use the existing module resolution logic.

As a curveball you could introduce an entirely new language construct to treat modules like providers and only support specific types using version constraints.

```terraform
terraform {
  required_modules {
    mymodule = {
      source  = "oci://example.com/terraform-modules/my-module"
      version = "~> 1.0"
    }
    myothermodule = {
      source  = "github.com/my-org/my-other-module"
      version = "~> 1.0"
    }
  }
}

module "my_module" {
  source = "module:mymodule"
}

module "my_other_module" {
  source = "module:myothermodule"
}
```

> In the meantime you the best option we have available is to use the two immutable source address types where a checksum is inherent in the source address: a git source with a specific commit id, or an OCI repository source with a specific image digest.

This would be a potential workaround for Git sourced modules if Dependabot supported tracking TF module digests the same way it tracks GitHub Actions. For OCI modules Dependabot would need to add explicit OpenTofu support on top of digest tracking.

**Comment by tmccombs** [comment]
> or have a content address (Git or OCI)

The problem with these is that a `ref` or `tag` isn't necessarily immutable. So if you have a hash in the lock file for, say, a git repo with `ref=main` or `ref=master`, or an OCI with `tag=latest`. Or even some other branch or moving tag, then an update would break modules downloads. 

At least in theory, it would be possible to use some convention for getting a list of versions from git and OCI sources. Ex,  look for tags/labels of the form `vX.Y.Z`. But it sounds like that is somewhat complicated with the existing use of the go-getter library. Ideally, the go-getter library would add support for that. Otherwise it would be necessary to fork the library, replace the library, or add separate functionality for getting available versions that works differently than the actual download.

I'd love it if at some point we had first-class version support for other module source types (git, OCI, maybe even local files), but I also think an MVP for this feature could be restricted to cases where we do version resolution (which as far as I am aware is currently only using a module registry).

**Comment by stevehipwell** [comment]
> The problem with these is that a ref or tag isn't necessarily immutable. So if you have a hash in the lock file for, say, a git repo with ref=main or ref=master, or an OCI with tag=latest. Or even some other branch or moving tag, then an update would break modules downloads.

@tmccombs this is what the lock file solves by storing the _resolved target_ when locking and then transforming the reference when fetching the module; `ref=main` -> `ref=45b23dee08af5e43a7fea6c4cf9c25ccf269ee113168c19722f87876677c5cb2`.

> At least in theory, it would be possible to use some convention for getting a list of versions from git and OCI sources. Ex, look for tags/labels of the form vX.Y.Z. But it sounds like that is somewhat complicated with the existing use of the go-getter library. Ideally, the go-getter library would add support for that. Otherwise it would be necessary to fork the library, replace the library, or add separate functionality for getting available versions that works differently than the actual download.
>
>I'd love it if at some point we had first-class version support for other module source types (git, OCI, maybe even local files), but I also think an MVP for this feature could be restricted to cases where we do version resolution (which as far as I am aware is currently only using a module registry).

I think a MVP could stick with the current language, supporting the registry with version constraints and Git & OCI by storing the _resolved target_. Resolving a Git or OCI target should happen before go-getter is involved, with creating the lockfile and checking the digests happening after.

**Comment by jack-bradshaw** [comment]
I have a request for this work to ensure static analysis of the module graph keeps working. In the present modules.json file, each entry has a key, source address, and directory, could these details please be retained in the new lockfile?

I ask because this file is presently the only way to traverse the graph of modules and associate each module block with its static address after variable expansion (without writing custom variable evaluation logic). I've looked into various other approaches using the public workflows, including `tofu show` and `tofu graph` commands, but none of them appear to provide this exact information (mapping between keys, dirs, and addresses), so static analysers are forced to use the modules.json file right now.

I think its reasonable that changes to the lock file will probably require changes to static analysers that parse it, but if these details were removed from the lock file entirely, it would make this kind of static dependency analysis impossible. That throws a fairly impassable spanner in the works for any form of dependency analysis which is critical for build-tools like Bazel.

Do you think retaining these details in the new lockfile would be feasible?

Side note: Perhaps a better long-term solution would be exposing the information with a separate command, since Tofu already has the information internally to create the JSON file. That might be more maintainable for everyone.
